mozjs-sys/src/jsgc.rs

```diff
@@ -2,91 +2,145 @@
-    }
+    type Vtable = ();
+    const VTABLE: Self::Vtable = ();
+    const KIND: JS::RootKind = JS::RootKind::Symbol;
 }
 
 impl RootKind for *mut JS::BigInt {
-    #[inline(always)]
-    fn rootKind() -> JS::RootKind {
-        JS::RootKind::BigInt
-    }
+    type Vtable = ();
+    const VTABLE: Self::Vtable = ();
+    const KIND: JS::RootKind = JS::RootKind::BigInt;
 }
 
 impl RootKind for *mut JSScript {
-    #[inline(always)]
-    fn rootKind() -> JS::RootKind {
-        JS::RootKind::Script
-    }
+    type Vtable = ();
+    const VTABLE: Self::Vtable = ();
+    const KIND: JS::RootKind = JS::RootKind::Script;
 }
 
 impl RootKind for jsid {
-    #[inline(always)]
-    fn rootKind() -> JS::RootKind {
-        JS::RootKind::Id
-    }
+    type Vtable = ();
+    const VTABLE: Self::Vtable = ();
+    const KIND: JS::RootKind = JS::RootKind::Id;
 }
 
 impl RootKind for JS::Value {
-    #[inline(always)]
-    fn rootKind() -> JS::RootKind {
-        JS::RootKind::Value
-    }
+    type Vtable = ();
+    const VTABLE: Self::Vtable = ();
+    const KIND: JS::RootKind = JS::RootKind::Value;
 }
 
 impl RootKind for JS::PropertyDescriptor {
-    #[inline(always)]
-    fn rootKind() -> JS::RootKind {
-        JS::RootKind::Traceable
+    type Vtable = *const RootedVFTable;
+    const VTABLE: Self::Vtable = &<Self as TraceableTrace>::VTABLE;
+    const KIND: JS::RootKind = JS::RootKind::Traceable;
+}
+
+/// A vtable for use in RootedTraceable<T>, which must be present for stack roots using
+/// RootKind::Traceable. The C++ tracing implementation uses a virtual trace function
+/// which is only present for C++ Rooted<T> values that use the Traceable root kind.
+#[repr(C)]
+pub struct RootedVFTable {
+    #[cfg(windows)]
+    pub padding: [usize; 1],
+    #[cfg(not(windows))]
+    pub padding: [usize; 2],
+    pub trace: unsafe extern "C" fn(this: *mut c_void, trc: *mut JSTracer, name: *const c_char),
+}
+
+impl RootedVFTable {
+    #[cfg(windows)]
+    pub const PADDING: [usize; 1] = [0];
+    #[cfg(not(windows))]
+    pub const PADDING: [usize; 2] = [0, 0];
+}
+
+/// `Rooted<T>` with a T that uses the Traceable RootKind uses dynamic dispatch on the C++ side
+/// for custom tracing. This trait provides trace logic via a vtable when creating a Rust instance
+/// of the object.
+pub unsafe trait TraceableTrace: Sized + RootKind {
+    const VTABLE: RootedVFTable = RootedVFTable {
+        padding: RootedVFTable::PADDING,
+        trace: Self::trace,
+    };
+
+    unsafe extern "C" fn trace(this: *mut c_void, trc: *mut JSTracer, _name: *const c_char) {
+        let rooted = this as *mut Rooted<Self>;
+        let rooted = rooted.as_mut().unwrap();
+        Self::do_trace(&mut rooted.ptr, trc);
     }
+
+    /// Used by `TraceableTrace` implementer to trace its contents.
+    /// Corresponds to virtual `trace` call in a `Rooted` that inherits from
+    /// StackRootedTraceableBase (C++).
+    fn do_trace(&mut self, trc: *mut JSTracer);
+}
```

Comment:
I think this trait is redundant as it's not used as bound but rather as namespace. It does provide wrapper for trace, but I think hops just complicates things further. I think this would be clearer:

```rust
#[repr(C)]
pub struct RootedVFTable {
    #[cfg(windows)]
    padding: [usize; 1],
    #[cfg(not(windows))]
    padding: [usize; 2],
    trace: unsafe extern "C" fn(this: *mut c_void, trc: *mut JSTracer, name: *const c_char),
}

impl RootedVFTable {
    const fn new(trace: unsafe extern "C" fn(this: *mut c_void, trc: *mut JSTracer, name: *const c_char)) -> Self {
        Self {
            padding: Self::PADDING,
            trace,
        }
    }
}

    unsafe extern "C" fn prop_desc_trace(this: *mut c_void, trc: *mut JSTracer, _name: *const c_char) {
        let rooted = this as *mut Rooted<Self>;
        let rooted = rooted.as_mut().unwrap();
        CallPropertyDescriptorTracer(trc, &mut rooted.ptr);
    }

const prop_desc_vtable = RootedVFTable::new(prop_desc_trace);

impl RootKind for JS::PropertyDescriptor {
    type Vtable = *const RootedVFTable;
    const VTABLE: Self::Vtable = &prop_desc_vtable;
    const KIND: JS::RootKind = JS::RootKind::Traceable;
}
```

Reply:
Yeah, I suppose that makes sense. I tried to use it as a bound like:
```rust
impl<T: TraceableTrace> for T {
    type Vtable = *const RootedVFTable;
    const VTABLE: Self::Vtable = &<Self as TraceableTrace>::VTABLE;
    const KIND: JS::RootKind = JS::RootKind::Traceable;
}
```
but rustc complained that the type of `&<Self as TraceableTrace>::VTABLE` was not the expected associated type.

Reply:
Actually, I got this to work in the latest commit and I'd prefer to keep TraceableTrace around for that reason. It makes it impossible to use the wrong vtable definition this way.

Reply:
That's perfect!